### DISCUSSION TITLE
Facing 403 Error in the middle of the tests while cypress running tests in dev environment

### DISCUSSION THREAD

**Discussion by kalaivanee** [cypress-io/cypress#14437] [Questions and Help]
Facing 403 Error while cypress running tests in dev environment 

![image](https://user-images.githubusercontent.com/77075297/103814496-c09afd80-502f-11eb-9d94-86fbafb81b3a.png)

**Comment by kalaivanee** [comment]
most of the tests are affected cos of this., If someone pour some insight on this, it will be helpfull

**Comment by bahmutov** [comment]
It is pretty hard to answer anything without a reproducible example or more details. Do you use cookies? Does it go from domain to domain? What is happening and how the requests are authenticated?

**Reply by kalaivanee** [reply]
Thanks for the reply.. I will post the code here.  It works on the single domain. 
/// <reference types="Cypress" />
import {
	navigateToEncounterNotes,
	createDraftNote,
} from '../../support/common/Notes/utils';
import {
	notesSelectors,
	patientToSelectInNotese2e,
} from '../../support/constants';
import { notesApiURLs } from '../../support/apiURLs';

describe('editNotes.spec.js', () => {
	context('Given that you are on the note type section with draft note', () => {
		before(() => {
			cy.server();
			cy.route(notesApiURLs.ENCOUNTER_NOTE_DETAIL).as('encounterNoteDetailApi');
			cy.loginAndWaitForPatientList();
			cy.selectPatient(patientToSelectInNotese2e);
			navigateToEncounterNotes();
			cy.navigateToNoteTemplate().then(createDraftNote);
		});

		it('should be able to select draft note for editing', () => {
			cy.wait('@encounterNoteApi');
			cy.get(notesSelectors.ENCOUNTER_NOTE_LIST)
				.find('tbody>tr>td')
				.eq(0)
				.should('contain', 'Draft')
				.click();
			cy.wait('@encounterNoteDetailApi');
			cy.get('[data-testid=note-creator-wrapper]').should('be.visible');
		});

		it('should be able to switch tabs in patient chart pane', () => {
			// Default tab should be notes
			cy.get(notesSelectors.NOTES_TAB)
				.invoke('attr', 'aria-selected')
				.should('contain', true);
			cy.verifyPatientChartTabs();
		});

		it('should be able to add content to note', () => {
			let contentLengthBefore;
			cy.get(notesSelectors.NOTE_TYPE_TEMPLATE_SECTIONS)
				.eq(0)
				.find(notesSelectors.CONTENT_EDITABLE)
				.invoke('text')
				.then(content => {
					contentLengthBefore = content.length;
				});

			cy.get(notesSelectors.NOTE_TYPE_TEMPLATE_SECTIONS)
				.eq(0)
				.find(notesSelectors.CONTENT_EDITABLE)
				.type('{enter}Additional content')
				.invoke('text')
				.then(content => {
					expect(content).to.have.length.greaterThan(contentLengthBefore);
				});
		});

		it('should be able to exit the note', () => {
			cy.server();
			cy.route(notesApiURLs.ENCOUNTER_NOTE).as('encounterNoteApi');
			cy.get(notesSelectors.NOTE_TYPE_TEMPLATE_EXIT_BUTTON).click();
			cy.wait('@encounterNoteApi');
			cy.get(notesSelectors.ENCOUNTER_NOTE_LIST).should('be.visible');
		});
	});
});


When it goes from 1 IT block to other block it throws 403 error, but when i have all the code in one IT block , it works absolutely fine.

**Reply by kalaivanee** [reply]
Now I have put all the code in one IT block, it runs absolutely fine without running into 403 error 
![image](https://user-images.githubusercontent.com/77075297/103915371-1ecfeb00-50d9-11eb-8853-234c691a7c47.png)

![image](https://user-images.githubusercontent.com/77075297/103915396-298a8000-50d9-11eb-996f-a5b2e7ec00ff.png)

**Comment by bahmutov** [comment]
Ok, you are logging in once using `before` hook https://docs.cypress.io/guides/core-concepts/writing-and-organizing-tests.html#Hooks

Every test resets cookies and other state data, thus you can either preserve cookies https://docs.cypress.io/faq/questions/using-cypress-faq.html#How-do-I-preserve-cookies-localStorage-in-between-my-tests or set them again (and any other session data) before each test. 

I think this documentation and linked video is a good answer https://docs.cypress.io/guides/references/best-practices.html#Organizing-Tests-Logging-In-Controlling-State

**Reply by kalaivanee** [reply]
Hi @bahmutov , i added some more additional details above. Did u got chance to look into it?

**Reply by bahmutov** [reply]
@kalaivanee I have already replied after looking at your code above

**Reply by kalaivanee** [reply]
@bahmutov : I have tried adding

 Cypress.Cookies.defaults({
  preserve: 'session_id'
}) to my login.js but still getting 403 error

**Reply by bahmutov** [reply]
Well, it is hard to say anything without running the application. We general logging in advice is in https://docs.cypress.io/examples/examples/recipes.html#Logging-In

**Comment by kalaivanee** [comment]
@bahmutov  : Thanks i was able to resolve the issue based on ur previous comment.
